File: src/main/java/leetcode/Problem1838.java
package leetcode;

import java.util.Arrays;

@Algorithms(Algorithm.SLIDING_WINDOW)
@BeatsPercent(88.51)
@TimeComplexity(worst = Complexity.LINEARITHMIC_N)
@SpaceComplexity(worst = Complexity.CONSTANT)
public class Problem1838 {
    public int maxFrequency(int[] nums, int k) {
        int n = nums.length;
        if (n < 2) {
            return n;
        }
        Arrays.sort(nums);
        int max = 0;
        long sum = 0;
        for (int i = 0, j = 0; j < n; j++) {
            long num = nums[j];
            sum += num;
            while (k + sum < (j - i + 1)*num) {
                sum -= nums[i++];
            }
            max = Math.max(max, j - i + 1);
        }
        return max;
    }
}

File: src/problem_1838.py
class Solution:
    def maxFrequency(self, nums: List[int], k: int) -> int:
        n, nums = len(nums), sorted(nums)
        lo = hi = n - 1
        c, res = k, 0
        while lo >= 0:
            while c >= 0:
                res = max(res, hi - lo + 1)
                lo -= 1
                if lo >= 0:
                    c -= nums[hi] - nums[lo]
                else:
                    break
            c += (hi - lo) * (nums[hi] - nums[hi - 1])
            hi -= 1
        return res
